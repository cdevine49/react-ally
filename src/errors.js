export const useSpecializedRole = (propName, componentName) =>
  `\`${componentName}\` doesn't accept a \`${propName}\` prop, if you need to use a specialized role, please use the \`Alert\`, \`Log\`, \`Status\`, \`Progressbar\`, \`Marguee\`, or \`Timer\` component.`;
export const invalidAriaRelevant = (value, propName, componentName) =>
  `Invalid value \`${value}\` for \`${propName}\` supplied to \`${componentName}\`. Validation failed.`;
export const blacklisted = (propName, componentName) =>
  `Invalid prop \`${propName}\` supplied to \`${componentName}\`. The default value for \`${propName}\` will not be overridden.`;
export const zeroToOne = (avn, propName, componentName) =>
  `Invalid value \`${avn}\` for \`${propName}\` supplied to \`${componentName}\`. Validation failed.`;
export const needsAriaLabelOrTitle = (ariaLabel, propName, componentName) =>
  `Invalid value \`${ariaLabel}\` for \`${propName}\` supplied to \`${componentName}\`.  ${componentName} must have either an AlertDialogLabel child or a valid aria-label`;

// Dialog
export const needsDialogLabel = componentName => `\`${componentName}\` expects a \`DialogLabel\` child`;
// AlertDialog
export const needsDialogDescription = componentName => `\`${componentName}\` expects a \`DialogDescription\` child`;
