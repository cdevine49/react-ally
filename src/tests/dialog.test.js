import React from "react";
import { cleanup, fireEvent, render } from "react-testing-library";
import {
  Dialog,
  DialogContent,
  DialogOpenButton,
  DialogCloseButton,
  LeastDangerousElement,
  FirstTabbableElement,
  DialogLabel,
  DialogDescription,
  LastTabbableElement
} from "../components/dialog";
import { needsDialogLabel, needsDialogDescription } from "../errors";

beforeEach(() => {
  console.error = jest.fn();
});
afterEach(cleanup);

test("default", () => {
  const component = (
    <Dialog>
      <DialogOpenButton>Open Me</DialogOpenButton>
      <DialogContent>
        <DialogLabel id="label-id">Content</DialogLabel>
      </DialogContent>
    </Dialog>
  );
  const { asFragment, unmount } = render(component, { container: document.body });
  expect(asFragment()).toMatchSnapshot();

  unmount();
});

test("initializeOpen prop", () => {
  const component = (
    <Dialog initializeOpen>
      <DialogOpenButton>Open Me</DialogOpenButton>
      <DialogContent>
        <DialogLabel id="label-id">Content</DialogLabel>
      </DialogContent>
    </Dialog>
  );
  const { asFragment, unmount } = render(component, { container: document.body });
  expect(asFragment()).toMatchSnapshot();

  unmount();
});

test("Dialog ignores children besides DialogOpenButton and DialogContent", () => {
  const component = (
    <Dialog initializeOpen>
      <div>Hello</div>
      <button>World</button>
    </Dialog>
  );
  const { asFragment, unmount } = render(component, { container: document.body });
  expect(asFragment()).toMatchSnapshot();

  unmount();
});

describe(`Opening a dialog`, () => {
  test("Clicking DialogOpenButton", () => {
    const component = (
      <Dialog>
        <DialogOpenButton>Open Me</DialogOpenButton>
        <DialogContent>
          <DialogLabel id="label-id">Content</DialogLabel>
        </DialogContent>
      </Dialog>
    );
    const { asFragment, getByText, unmount } = render(component, { container: document.body });

    fireEvent.click(getByText("Open Me"));

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  test(`Focus goes to LeastDangerousElement`, () => {
    const onFocusFirstTabbable = jest.fn();
    const onFocusLeastDangerous = jest.fn();
    const onBlurLeastDangerous = jest.fn();
    const component = (
      <Dialog>
        <DialogOpenButton>Open Me</DialogOpenButton>
        <DialogContent>
          <DialogLabel id="label-id">Header</DialogLabel>
          <FirstTabbableElement>
            <button onFocus={onFocusFirstTabbable}>First Tabbable</button>
          </FirstTabbableElement>
          <LeastDangerousElement>
            <button onFocus={onFocusLeastDangerous} onBlur={onBlurLeastDangerous}>
              Least Dangerous
            </button>
          </LeastDangerousElement>
        </DialogContent>
      </Dialog>
    );
    const { getByText, rerender, unmount } = render(component, { container: document.body });
    fireEvent.click(getByText("Open Me"));
    rerender(component);

    expect(onFocusFirstTabbable).not.toHaveBeenCalled();
    expect(onFocusLeastDangerous).toHaveBeenCalled();
    expect(onBlurLeastDangerous).not.toHaveBeenCalled();

    unmount();
  });

  test(`Focus goes to FirstTabbableElement if no LeastDangerousElement`, () => {
    const onFocusFirstTabbable = jest.fn();
    const onFocusOther = jest.fn();
    const onBlurFirstTabbable = jest.fn();
    const component = (
      <Dialog>
        <DialogOpenButton>Open Me</DialogOpenButton>
        <DialogContent>
          <DialogLabel id="label-id">Header</DialogLabel>
          <button onFocus={onFocusOther}>I dont receive focus</button>
          <FirstTabbableElement>
            <button onFocus={onFocusFirstTabbable} onBlur={onBlurFirstTabbable}>
              First Tabbable
            </button>
          </FirstTabbableElement>
          <button onFocus={onFocusOther}>I dont receive focus</button>
        </DialogContent>
      </Dialog>
    );
    const { getByText, rerender, unmount } = render(component, { container: document.body });
    fireEvent.click(getByText("Open Me"));
    rerender(component);

    expect(onFocusFirstTabbable).toHaveBeenCalled();
    expect(onBlurFirstTabbable).not.toHaveBeenCalled();
    expect(onFocusOther).not.toHaveBeenCalled();

    unmount();
  });

  test(`Sets modalRoot's siblings to aria-hidden`, () => {
    const component = (
      <div>
        <Dialog>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent>
            <DialogLabel id="label-id">Content</DialogLabel>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, getByText, rerender, unmount } = render(component, { container: document.body });
    const sibling = document.createElement("div");
    document.body.appendChild(sibling);
    const hiddenSibling = document.createElement("div");
    hiddenSibling.setAttribute("aria-hidden", true);
    document.body.appendChild(hiddenSibling);
    fireEvent.click(getByText("Open Me"));
    rerender(component);
    expect(asFragment()).toMatchSnapshot();

    unmount();
  });
});

describe("Closing the dialog", () => {
  test("Clicking DialogCloseButton", () => {
    const component = (
      <div>
        <Dialog initializeOpen>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent>
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogCloseButton>Close Me</DialogCloseButton>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, getByText, rerender, unmount } = render(component, { container: document.body });

    fireEvent.click(getByText("Close Me"));
    rerender(component);

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  test("Click the overlay", () => {
    const component = (
      <div>
        <Dialog initializeOpen>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent overlayProps={{ "data-testid": "dialog-overlay" }}>
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogCloseButton>Close Me</DialogCloseButton>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, getByTestId, rerender, unmount } = render(component, { container: document.body });

    fireEvent.click(getByTestId("dialog-overlay"));
    rerender(component);

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  test("Click the dialog", () => {
    const component = (
      <div>
        <Dialog initializeOpen>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent data-testid="dialog">
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogCloseButton>Close Me</DialogCloseButton>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, getByTestId, rerender, unmount } = render(component, { container: document.body });

    fireEvent.click(getByTestId("dialog"));
    rerender(component);

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  test("Escape key", () => {
    const component = (
      <div>
        <Dialog initializeOpen>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent>
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogCloseButton>Close Me</DialogCloseButton>
            <button>Focus Me</button>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, getByText, rerender, unmount } = render(component, { container: document.body });

    fireEvent.click(getByText("Focus Me"));
    rerender(component);
    fireEvent.keyDown(getByText("Focus Me"), {
      key: "Escape",
      keyCode: 27
    });
    rerender(component);

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  test("Focus returns to open button", () => {
    const onFocusOpenButton = jest.fn();
    const onBlurOpenButton = jest.fn();
    const component = (
      <Dialog initializeOpen>
        <DialogOpenButton onFocus={onFocusOpenButton} onBlur={onBlurOpenButton}>
          Open Me
        </DialogOpenButton>
        <DialogContent>
          <DialogLabel id="label-id">Header</DialogLabel>
          <DialogCloseButton>Close Me</DialogCloseButton>
        </DialogContent>
      </Dialog>
    );
    const { getByText, rerender, unmount } = render(component, { container: document.body });
    fireEvent.click(getByText("Close Me"));
    rerender(component);

    expect(onFocusOpenButton).toHaveBeenCalled();
    expect(onBlurOpenButton).not.toHaveBeenCalled();

    unmount();
  });

  test("ModalRoot siblings return to original aria-hidden states", () => {
    const component = (
      <div>
        <Dialog initializeOpen>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent>
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogCloseButton>Close Me</DialogCloseButton>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, getByText, rerender, unmount } = render(component, { container: document.body });

    const sibling = document.createElement("div");
    document.body.appendChild(sibling);
    const hiddenSibling = document.createElement("div");
    hiddenSibling.setAttribute("aria-hidden", true);
    document.body.appendChild(hiddenSibling);

    rerender(component);
    fireEvent.click(getByText("Close Me"));
    rerender(component);

    expect(asFragment()).toMatchSnapshot();

    unmount();
  });
});

describe("DialogContent", () => {
  test("expects a DialogLabel child", () => {
    const component = <DialogContent>Content</DialogContent>;
    render(component);
    expect(console.error).toHaveBeenCalledWith(`Warning: Failed prop type: ${needsDialogLabel("DialogContent")}
    in DialogContent`);
  });

  test(`aria-labelledby DialogLabel's id`, () => {
    const component = (
      <DialogContent>
        <DialogLabel id="header-id" />
      </DialogContent>
    );
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test(`aria-describedby DialogDescription's id`, () => {
    const component = (
      <DialogContent>
        <DialogLabel id="header-id" />
        <DialogDescription id="description-id" />
      </DialogContent>
    );
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("DialogLabel", () => {
  test("id is required", () => {
    const component = <DialogLabel />;
    render(component);
    expect(console.error)
      .toHaveBeenCalledWith(`Warning: Failed prop type: The prop \`id\` is marked as required in \`DialogLabel\`, but its value is \`undefined\`.
    in DialogLabel`);
  });
});

describe("DialogDescription", () => {
  test("id is required", () => {
    const component = <DialogDescription />;
    render(component);
    expect(console.error)
      .toHaveBeenCalledWith(`Warning: Failed prop type: The prop \`id\` is marked as required in \`DialogDescription\`, but its value is \`undefined\`.
    in DialogDescription`);
  });
});

describe("AlertDialog", () => {
  test("DialogDescription is required", () => {
    const component = (
      <div>
        <Dialog>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent role="alertdialog">
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogCloseButton>Close Me</DialogCloseButton>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { unmount } = render(component);

    expect(console.error).toHaveBeenCalledWith(`Warning: Failed prop type: ${needsDialogDescription("DialogContent")}
    in DialogContent`);

    unmount();
  });

  test(`has correct role`, () => {
    const component = (
      <div>
        <Dialog initializeOpen>
          <DialogOpenButton>Open Me</DialogOpenButton>
          <DialogContent role="alertdialog">
            <DialogLabel id="label-id">Header</DialogLabel>
            <DialogDescription id="alert-desc">Don't Delete That Just Yet</DialogDescription>
            <DialogCloseButton>Close Me</DialogCloseButton>
          </DialogContent>
        </Dialog>
      </div>
    );
    const { asFragment, unmount } = render(component, { container: document.body });

    expect(asFragment()).toMatchSnapshot();

    unmount();
  });
});

test("Nested Dialog Modals", () => {
  const component = (
    <div>
      <Dialog>
        <DialogOpenButton>Open First</DialogOpenButton>
        <DialogContent>
          <DialogLabel id="label-id">First Header</DialogLabel>
          <Dialog>
            <DialogOpenButton>Open Second</DialogOpenButton>
            <DialogContent>
              <DialogLabel id="label-id">Second Header</DialogLabel>
              <DialogCloseButton>Close Second</DialogCloseButton>
            </DialogContent>
          </Dialog>
          <DialogCloseButton>Close First</DialogCloseButton>
        </DialogContent>
      </Dialog>
    </div>
  );

  const { asFragment, getByText, rerender, unmount } = render(component, { container: document.body });

  expect(asFragment()).toMatchSnapshot();

  fireEvent.click(getByText("Open First"));
  rerender(component);

  expect(asFragment()).toMatchSnapshot();

  fireEvent.click(getByText("Open Second"));
  rerender(component);

  expect(asFragment()).toMatchSnapshot();

  unmount();
});
