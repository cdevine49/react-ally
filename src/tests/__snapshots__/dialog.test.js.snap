// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertDialog has correct role 1`] = `
<DocumentFragment>
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-describedby="alert-desc"
        aria-labelledby="label-id"
        aria-modal="true"
        role="alertdialog"
      >
        <h2
          id="label-id"
        >
          Header
        </h2>
        <div
          id="alert-desc"
        >
          Don't Delete That Just Yet
        </div>
        <button>
          Close Me
        </button>
      </div>
    </div>
  </div>
  <div>
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
</DocumentFragment>
`;

exports[`Closing the dialog Click the dialog 1`] = `
<DocumentFragment>
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        data-testid="dialog"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          Header
        </h2>
        <button>
          Close Me
        </button>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
  >
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
</DocumentFragment>
`;

exports[`Closing the dialog Click the overlay 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
  >
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
</DocumentFragment>
`;

exports[`Closing the dialog Clicking DialogCloseButton 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
  >
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
</DocumentFragment>
`;

exports[`Closing the dialog Escape key 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
  >
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
</DocumentFragment>
`;

exports[`Closing the dialog ModalRoot siblings return to original aria-hidden states 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
  >
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
  <div
    aria-hidden="false"
  />
  <div
    aria-hidden="true"
  />
</DocumentFragment>
`;

exports[`Dialog ignores children besides DialogOpenButton and DialogContent 1`] = `<DocumentFragment />`;

exports[`DialogContent aria-describedby DialogDescription's id 1`] = `
<DocumentFragment>
  <div
    style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
  >
    <div
      aria-describedby="description-id"
      aria-labelledby="header-id"
      aria-modal="true"
      role="dialog"
    >
      <h2
        id="header-id"
      />
      <div
        id="description-id"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`DialogContent aria-labelledby DialogLabel's id 1`] = `
<DocumentFragment>
  <div
    style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
  >
    <div
      aria-labelledby="header-id"
      aria-modal="true"
      role="dialog"
    >
      <h2
        id="header-id"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Nested Dialog Modals 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-disabled="false"
    >
      Open First
    </button>
  </div>
</DocumentFragment>
`;

exports[`Nested Dialog Modals 2`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
  >
    <button
      aria-disabled="false"
    >
      Open First
    </button>
  </div>
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          First Header
        </h2>
        <button
          aria-disabled="false"
        >
          Open Second
        </button>
        <button>
          Close First
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Nested Dialog Modals 3`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
  >
    <button
      aria-disabled="false"
    >
      Open First
    </button>
  </div>
  <div
    aria-hidden="true"
  >
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          First Header
        </h2>
        <button
          aria-disabled="false"
        >
          Open Second
        </button>
        <button>
          Close First
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          Second Header
        </h2>
        <button>
          Close Second
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Opening a dialog Clicking DialogOpenButton 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
  >
    Open Me
  </button>
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          Content
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Opening a dialog Sets modalRoot's siblings to aria-hidden 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
  >
    <button
      aria-disabled="false"
    >
      Open Me
    </button>
  </div>
  <div
    aria-hidden="true"
  />
  <div
    aria-hidden="true"
  />
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          Content
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`default 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
  >
    Open Me
  </button>
</DocumentFragment>
`;

exports[`initializeOpen prop 1`] = `
<DocumentFragment>
  <div>
    <div
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-labelledby="label-id"
        aria-modal="true"
        role="dialog"
      >
        <h2
          id="label-id"
        >
          Content
        </h2>
      </div>
    </div>
  </div>
  <button
    aria-disabled="false"
  >
    Open Me
  </button>
</DocumentFragment>
`;
